apps/src/javalab/redux/editorRedux.ts

```diff
@@ -0,0 +1,320 @@
+        return {payload: {filename, isValidation}};
+      }
+    },
+    sourceFileOrderUpdated(state) {
+      const sources = _.cloneDeep(state.sources);
+      const updatedSources = updateAllSourceFileOrders(
+        sources,
+        state.fileMetadata,
+        state.orderedTabKeys
+      );
+      state.sources = updatedSources;
+    },
+    removeFile(state, action: PayloadAction<string>) {
+      const filename = action.payload;
+      const newSources = {...state.sources};
+      delete newSources[filename];
+      state.sources = newSources;
+    },
+    openEditorDialog(state, action: PayloadAction<JavlabEditorDialogOptions>) {
+      if (Object.values(JavalabEditorDialog).includes(action.payload)) {
+        state.editorOpenDialogName = action.payload;
+      }
+    },
+    closeEditorDialog(state) {
+      state.editorOpenDialogName = null;
+    },
+    setEditTabKey(state, action: PayloadAction<string>) {
+      state.editTabKey = action.payload;
+    },
+    setActiveTabKey(state, action: PayloadAction<string>) {
+      state.activeTabKey = action.payload;
+    },
+    setFileMetadata(state, action: PayloadAction<FileMetadata>) {
+      state.fileMetadata = action.payload;
+    },
+    setOrderedTabKeys(state, action: PayloadAction<string[]>) {
+      state.orderedTabKeys = action.payload;
+    },
+    setAllEditorMetadata: {
+      reducer(
+        state,
+        action: PayloadAction<{
+          fileMetadata: FileMetadata;
+          orderedTabKeys: string[];
+          activeTabKey: string;
+          lastTabKeyIndex: number;
+        }>
+      ) {
+        state.fileMetadata = action.payload.fileMetadata;
+        state.orderedTabKeys = action.payload.orderedTabKeys;
+        state.activeTabKey = action.payload.activeTabKey;
+        state.lastTabKeyIndex =
+          action.payload.lastTabKeyIndex || state.lastTabKeyIndex;
+      },
+      prepare(
+        fileMetadata: FileMetadata,
+        orderedTabKeys: string[],
+        activeTabKey: string,
+        lastTabKeyIndex: number
+      ) {
+        return {
+          payload: {fileMetadata, orderedTabKeys, activeTabKey, lastTabKeyIndex}
+        };
+      }
+    },
+    setNewFileError(state, action: PayloadAction<string>) {
+      state.newFileError = action.payload;
+    },
+    clearNewFileError(state) {
+      state.newFileError = null;
+    },
+    setRenameFileError(state, action: PayloadAction<string>) {
+      state.renameFileError = action.payload;
+    },
+    clearRenameFileError(state) {
+      state.renameFileError = null;
+    }
+  }
+});
+
+// Selectors
+// We don't have a type for the entire redux store yet, so we need to use any here.
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const selectSources = (state: any) => state.javalabEditor.sources;
+
+export const getSources = createSelector(selectSources, sources => {
+  const result: Sources = {};
+  for (const key in sources) {
+    if (!sources[key].isValidation) {
```

Comment:
I know this is old code that's been converted, but looking at this again, do you know why we have selectors to filter out validation or non-validation code from `javalabEditor.sources` but also have a separate `javalabEditor.validation` value?

Reply:
We only use `javalabEditor.validation` to check for duplicate file names when not in start mode (so when the validation files are hidden). The naming could be better here, we are really only tracking the filenames. `getSources` and `getValidation` are used for saving code, either in start mode (we save both), or normal mode (only save sources).